Source/NSPathControl.m

```diff
@@ -0,0 +1,396 @@
+                               forLocal: (BOOL)local
+{
+  if (local)
+    {
+      _localMask = mask;
+    }
+  else
+    {
+      _remoteMask = mask;
+    }
+}
+
+- (NSMenu *) menu
+{
+  return [super menu];
+}
+
+- (void) setMenu: (NSMenu *)menu
+{
+  [super setMenu: menu];
+}
+
+- (NSArray *) allowedTypes;
+{
+  return [_cell allowedTypes];
+}
+
+- (void) setAllowedTypes: (NSArray *)allowedTypes
+{
+  [_cell setAllowedTypes: allowedTypes];
+  [self registerForDraggedTypes: allowedTypes];
+}
+
+- (NSPathControlItem *) clickedPathItem
+{
+  return nil;
+}
+
+- (NSArray *) pathItems
+{
+  return _pathItems;
+}
+
+- (void) setPathItems: (NSArray *)items
+{
+  ASSIGNCOPY(_pathItems, items);
+  [self setNeedsDisplay];
+}
+
+- (NSAttributedString *) placeholderAttributedString
+{
+  return [_cell placeholderAttributedString];
+}
+
+- (void) setPlaceholderAttributedString: (NSAttributedString *)string
+{
+  [_cell setPlaceholderAttributedString: string];
+  [self setNeedsDisplay];
+}
+
+- (NSString *) placeholderString
+{
+  return [_cell placeholderString];
+}
+
+- (void) setPlaceholderString: (NSString *)string
+{
+  [_cell setPlaceholderString: string];
+  [self setNeedsDisplay];
+}
+
+- (NSColor *) backgroundColor
+{
+  return _backgroundColor;
+}
+
+- (void) setBackgroundColor: (NSColor *)color
+{
+  ASSIGN(_backgroundColor, color);
+  [self setNeedsDisplay];
+}
+
+- (void) setAction: (SEL)action
+{
+  _action = action;
+}
+
+- (SEL) action
+{
+  return _action;
+}
+
+- (void) setTarget: (id)target
+{
+  _target = target;
+}
+
+- (id) target
+{
+  return _target;
+}
+
+- (void) _doMenuAction: (id)sender
+{
+  if (_action)
+    {
+      [self sendAction: _action
+                    to: _target];
+    }
+
+  [[sender menu] close];
+}
+
+- (void) _doChooseMenuAction: (id)sender
+{
+  NSOpenPanel *op = [NSOpenPanel openPanel];
+  int result = 0;
+  
+  [op setAllowsMultipleSelection: NO];
+  [op setCanChooseFiles: YES];
+  [op setCanChooseDirectories: YES];
+  
+  result = [op runModalForDirectory: nil
+                               file: nil
+                              types: nil];
+  if (result == NSOKButton)
+    {
+      NSArray *urls = [op URLs];
+      NSURL *url = [urls objectAtIndex: 0];
+      [self setURL: url];
+    }
+
+  [[sender menu] close];
+}
+
+- (void) mouseDown: (NSEvent *)event
+{
+  if (![self isEnabled])
+    {
+      [super mouseDown: event];
+      return;
+    }
+
+  if ([self pathStyle] == NSPathStylePopUp)
+    {
+      NSPathCell *acell = (NSPathCell *)[self cell];
+      NSArray *array = [acell pathComponentCells];
+      NSMenu *menu = [[NSMenu alloc] initWithTitle: @"Select File"];
+      NSPathComponentCell *c = nil;
+      NSEnumerator *en = [array objectEnumerator];
+      
+      while((c = [en nextObject]) != nil)
+        {
+          NSURL *u = [c URL];
+          NSString *s = [[u path] lastPathComponent];
+          NSMenuItem *i = [[NSMenuItem alloc] init];
```

Comment:
The menu item is never released. Using one letter variable names is very helpful.